### PR TITLE
[release/2.1] Fix HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP setup

### PR DESCRIPTION
**Pull request by edenlee1212**
Because of build break, HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMA is not determined correctly
```
Performing Test HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP - Failed

Performing C SOURCE FILE Test HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP failed with the following output:
Change Dir: /opt/code/bin/obj/Linux.x64.Release/native/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_2967b/fast"
/usr/bin/make -f CMakeFiles/cmTC_2967b.dir/build.make CMakeFiles/cmTC_2967b.dir/build
make[1]: Entering directory '/opt/code/bin/obj/Linux.x64.Release/native/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_2967b.dir/src.c.o
/usr/bin/clang-3.9    -std=gnu99 -D_GNU_SOURCE -DHAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP -Werror   -o CMakeFiles/cmTC_2967b.dir/src.c.o   -c /opt/code/bin/obj/Linux.x64.Release/native/CMakeFiles/CMakeTmp/src.c
/opt/code/bin/obj/Linux.x64.Release/native/CMakeFiles/CMakeTmp/src.c:17:18: error: use of undeclared identifier 'nullptr'
        if (mmap(nullptr, 1, PROT_EXEC, MAP_PRIVATE, fd, 0) == MAP_FAILED)
                 ^
1 error generated.
CMakeFiles/cmTC_2967b.dir/build.make:65: recipe for target 'CMakeFiles/cmTC_2967b.dir/src.c.o' failed
make[1]: *** [CMakeFiles/cmTC_2967b.dir/src.c.o] Error 1
make[1]: Leaving directory '/opt/code/bin/obj/Linux.x64.Release/native/CMakeFiles/CMakeTmp'
Makefile:126: recipe for target 'cmTC_2967b/fast' failed
make: *** [cmTC_2967b/fast] Error 2

Return value: 1
Source file was:

    #include <sys/mman.h>
    #include <fcntl.h>
    #include <unistd.h>

    int main()
    {
        int fd = shm_open("/corefx_configure_shm_open", O_CREAT | O_RDWR, 0777);
        if (fd == -1)
            return -1;

        shm_unlink("/corefx_configure_shm_open");

        // NOTE: PROT_EXEC and MAP_PRIVATE don't work well with shm_open
        //       on at least the current version of Mac OS X

        if (mmap(nullptr, 1, PROT_EXEC, MAP_PRIVATE, fd, 0) == MAP_FAILED)
            return -1;

        return 0;
    }
```

### CODE REVIEW

**Comment by danmosemsft** [comment]
This isn't the first time I've of those little cmake test compilations was quietly failing for a reason other than the absence of the thing it was testing for. Is there a log where you can do a quick scan for any other unexpected failure reasons?

File: src/Native/Unix/configure.cmake [hunk 1]
```diff
@@ -444,7 +444,7 @@ check_c_source_compiles(
 
 set (CMAKE_REQUIRED_FLAGS ${PREVIOUS_CMAKE_REQUIRED_FLAGS})
 
-check_c_source_runs(
+check_cxx_source_runs(
```
[the remark this answers is not in the document]
Comment (edenlee1212) [reply src/Native/Unix/configure.cmake]: OK, I'll do that.

**Comment by edenlee1212** [comment]
@bartonjs I change nullptr to NULL instead of using check_c_source_runs.

**Comment by edenlee1212** [comment]
@dotnet-bot help

**Comment by edenlee1212** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by bartonjs** [comment]
@danmosemsft I assume this is should stay in a holding position, since it's the release branch.  Ja?

**Comment by danmosemsft** [comment]
Yes branch won't be open for a couple weeks I expect. Please circle the airport.

**Comment by danmosemsft** [comment]
Since this isn't breaking the actual official build we should have an issue with a brief shiproom template also.

**Comment by karelz** [comment]
Who can put together the shiproom template? (@danmosemsft @bartonjs @wtgodbe)

**Comment by wtgodbe** [comment]
@edenlee1212 is there already a github issue open for this? I can put together the shiproom template

**Comment by karelz** [comment]
I don't think there is GH issue. Please create one if needed (the new process puts templates on PRs, so it may not be needed - I am not 100% sure ...)

**Comment by wtgodbe** [comment]
#### Description
This CMake test is designed to check for the presence or absence of a feature on the current platform. It has been failing with a compilation error, and therefore failing on every platform, regardless of the presence or absence of the feature it's testing for. We should fix the code for this test so that we can get back the coverage that it's supposed to provide.
 
#### Customer Impact
The product is not working exactly as intended on platforms where we have an SHM-open that works well enough with MMAP

#### Regression?
No 

#### Risk
No additional risk beyond what the current status quo is - the test is either fixed, or it continues to fail.

Issue - https://github.com/dotnet/corefx/issues/31784

**Comment by wtgodbe** [comment]
Request has been sent to shiproom

**Comment by karelz** [comment]
@wtgodbe what was the shiproom outcome?

**Comment by wtgodbe** [comment]
Shiproom decided to sit on this given the lack of evidence of concrete brokenness. @danmosemsft when would be a good time to bring this up again?

**Comment by karelz** [comment]
@wtgodbe if it was rejected, let's close it. If there were follow up questions, let's make sure we are following up on them.

**Comment by wtgodbe** [comment]
There was a discussion over email where shiproom concluded that this had more risk (churn) than potential reward, I think closing is the right thing.

**Comment by wtgodbe** [comment]
@edenlee1212 thanks for this contribution - we've decided in shiproom that, since this doesn't affect the build of the product, it isn't worth the risk to take this into our Release branches. That being said, it's possible we missed some important consequence of this change in our conversations, so please feel free to raise any concerns if you feel that we should reconsider.

**Pull request closed without merging**